sdk/metric/controller/basic/controller.go

```diff
@@ -0,0 +1,300 @@
+// options (including an optional Exporter) to configure an metric
+// export pipeline.
+func New(checkpointer export.Checkpointer, opts ...Option) *Controller {
+	c := &Config{
+		CollectPeriod:  DefaultPeriod,
+		CollectTimeout: DefaultPeriod,
+		ExportTimeout:  DefaultPeriod,
+	}
+	for _, opt := range opts {
+		opt.Apply(c)
+	}
+
+	impl := sdk.NewAccumulator(
+		checkpointer,
+		c.Resource,
+	)
+	return &Controller{
+		provider:      registry.NewMeterProvider(impl),
+		accumulator:   impl,
+		checkpointer:  checkpointer,
+		exporter:      c.Exporter,
+		stopCh:        make(chan struct{}),
+		period:        c.CollectPeriod,
+		timeout:       c.CollectTimeout,
+		exportTimeout: c.ExportTimeout,
+		clock:         controllerTime.RealClock{},
+	}
+}
+
+// SetClock supports setting a mock clock for testing.  This must be
+// called before Start().
+func (c *Controller) SetClock(clock controllerTime.Clock) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	c.clock = clock
+}
+
+// MeterProvider returns a MeterProvider instance for this controller.
+func (c *Controller) MeterProvider() metric.MeterProvider {
+	return c.provider
+}
+
+// Start begins a ticker that periodically collects and exports
+// metrics with the configured interval.  This is required for calling
+// a configured Exporter (see WithExporter) and is otherwise optional
+// when only pulling metric data.
+//
+// The passed context is passed to Collect() and subsequently to
+// asynchronous instrument callbacks.  Returns an error when the
+// controller was already started.
+//
+// Note that it is not necessary to Start a controller when only
+// pulling data; use the Collect() and ForEach() methods directly in
+// this case.
+func (c *Controller) Start(ctx context.Context) error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	if c.ticker != nil {
+		return ErrControllerStarted
+	}
+
+	c.wg.Add(2)
+	c.ticker = c.clock.Ticker(c.period)
+	go c.runTicker(ctx, c.stopCh)
+	return nil
+}
+
+// Stop waits for the background goroutine to return and then collects
+// and exports metrics one last time before returning.  The passed
+// context is passed to Collect() and subsequently to asynchronous
+// instruments.
+func (c *Controller) Stop(ctx context.Context) error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	if c.stopCh == nil {
+		return nil
+	}
+
+	close(c.stopCh)
+	c.stopCh = nil
+	c.wg.Done()
+	c.wg.Wait()
+	c.ticker.Stop()
+
+	return c.collect(ctx, nil)
+}
+
+// runTicker collection on ticker events until the stop channel is closed.
+func (c *Controller) runTicker(ctx context.Context, stopCh chan struct{}) {
+	defer c.wg.Done()
+	for {
+		select {
+		case <-stopCh:
+			return
+		case <-c.ticker.C():
+			if err := c.collect(ctx, stopCh); err != nil {
+				otel.Handle(err)
+			}
+		}
+	}
+}
+
+// collect computes a checkpoint and optionally exports it.
+func (c *Controller) collect(ctx context.Context, stopCh chan struct{}) error {
+	if err := c.checkpoint(ctx, func() bool {
+		return true
```

Comment:
You probably want to call `select` on `stopCh` here.

Reply:
I've removed the parameter. now `stopCh` is only used for signaling the background thread to stop ticking.